proj/hps_accel/gateware/math.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2021 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from nmigen import Cat, Mux, Signal, unsigned, signed
+from nmigen.hdl.dsl import Module
+from nmigen_cfu import InstructionBase, SimpleElaboratable
+
+from .constants import Constants
+from .stream import Endpoint, connect
+
+SRDHM_INPUT_LAYOUT = [
+    ('a', signed(32)),
+    ('b', signed(32)),
+]
+
+
+class SaturatingRoundingDoubleHighMul(SimpleElaboratable):
+    """Performs an SRDHM operation.
+
+    This is approximately a 32x32 mulitiplication where only the high 32
+    bits are returned. It actually returns double the value, which is
+    possible because both inputs are signed and so the raw result of the
+    multiplication is only a 63 bit number.
+
+    Only tested for
+
+    * 'a' in the range +/- (64 * 16 * 128) (i.e 18bits including sign).
+      This is the range required for processing 4x4 convs with depths
+      up to 64.
+    * 'b' in the range 0x4000_0000 to 0x7000_0000 as this appears to be
+      the range used the hps model.
+
+    Attributes
+    ----------
+
+    input: Stream(SRDHM_INPUT_LAYOUT), in
+      The calculation to perform.
+
+    output: Stream(signed(32)), out
+      The result.
+
+    This pipeline neither provides nor respects back pressure.
+    """
+    PIPELINE_CYCLES = 3
+
+    def __init__(self):
+        self.input = Endpoint(SRDHM_INPUT_LAYOUT)
+        self.output = Endpoint(signed(32))
+
+    def delay(self, m, cycles, sig):
+        sr = Signal(cycles)
+        m.d.sync += sr.eq(Cat(sig, sr[:-1]))
+        return sr[-1]
+
+    def elab(self, m):
+        # Input is always ready
+        m.d.comb += self.input.ready.eq(1)
+
+        # Output's valid is input's valid, but delayed
+        m.d.comb += self.output.valid.eq(
+            self.delay(m, self.PIPELINE_CYCLES, self.input.valid))
+
+        # Cycle 0: register inputs
+        a = self.input.payload.a
+        reg_a = Signal(signed(32))
+        reg_b = Signal(signed(32))
+        m.d.sync += reg_a.eq(Mux(a >= 0, a, -a))
+        m.d.sync += reg_b.eq(self.input.payload.b)
+
+        # Cycle 1: multiply to register
+        # both operands are positive, so result always positive
```

Comment:
Should the `reg_ab` (and `reg_a` and `reg_b`) signals not be unsigned then? Or do we still need a sign bit for some reason?

Reply:
I think you're right, but the least significant bit of the result is dependent on a some math that I don't fully understand. I'd be 100% OK with changes to simplify this code further.